src/passes/StringLowering.cpp

```diff
@@ -240,8 +245,24 @@ struct StringLowering : public StringGathering {
     updates[HeapType::stringview_wtf16] = HeapType::ext;
     updates[HeapType::stringview_iter] = HeapType::ext;
 
+    // The module may have its own array16 type inside a big rec group, but
+    // imported strings expects that type in its own rec group as part of the
+    // ABI. Fix that up here. (This is valid to do as this type has no sub- or
+    // super-types anyhow; it is "plain old data" for communicating with the
+    // outside.)
+    auto allTypes = ModuleUtils::collectHeapTypes(*module);
+    auto array16 = nullArray16.getHeapType();
+    auto array16Element = array16.getArray().element;
+    for (auto type : allTypes) {
+      if (type.isArray()) {
+        if (type.getArray().element == array16Element) {
```

Comment:
This should also check that the finality and supertype of the replaced type are as expected.

We also might want some kind of strategy for handling the case when there are multiple types that could be replaced. If different i16 array types are differentiated by e.g. a cast, it would change program behavior to replace them both with the same type. Not sure how much we care about perfect correctness here, though.

If we do care about correctness, erroring out when there are multiple candidates could work. For better precision, we could restrict the search for candidates to places where we know we will need the isolated type.

Reply:
Good point, added a check for finality and supertype and testing for that. (Testing the supertype is not possible, though, as it won't validate - these are immutable arrays.)

I don't think we care too much about full correctness here - this is a limited pass that we'll adjust as needed.

For now I think we should best allow multiple array types, actually, as I can see use cases for them even if none exist atm: Imagine that someone wasm-merges two modules, then they could have separate array16s in their two big rec groups, but they should both be turned into the single one for external use. You're right that comparing those types would then be different, but that seems less of a priority to handle than being able to build such a module.